libantworld/render_target_input.cc

```diff
@@ -0,0 +1,38 @@
+#include "render_target_input.h"
+
+// Libantworld includes
+#include "render_target.h"
+
+//----------------------------------------------------------------------------
+// BoBRobotics::AntWorld::RenderTargetInput
+//----------------------------------------------------------------------------
+namespace BoBRobotics
+{
+namespace AntWorld
+{
+cv::Size RenderTargetInput::getOutputSize() const
+{
+    return cv::Size(m_RenderTarget.getWidth(), m_RenderTarget.getHeight());
+}
+//----------------------------------------------------------------------------
+bool RenderTargetInput::readFrame(cv::Mat &outFrame)
+{
+    // Make sure frame is of right size and type
+    outFrame.create(m_RenderTarget.getHeight(), m_RenderTarget.getWidth(), CV_8UC3);
+
+    // Bind render target texture
+    glBindTexture(GL_TEXTURE_2D, m_RenderTarget.getTexture());
+
+    // Read texture into outframe
+    glGetTexImage(GL_TEXTURE_2D, 0, GL_BGR, GL_UNSIGNED_BYTE, outFrame.data);
+
+    // Flip image vertically
+    cv::flip(outFrame, outFrame, 0);
+
+    // Unbind texture
+    glBindTexture(GL_TEXTURE_2D, 0);
+
+    return true;
```

Comment:
Does this implementation differ from that in ``Video::OpenGL``? Could they be amalgamated (e.g. with inheritance)?

Reply:
Yes - ``Video::OpenGL`` reads data from the 'display framebuffer' which is basically the same data the OS uses to draw to the screen. ``RenderTarget`` lets you not render to that and instead render to an offscreen buffer and ``RenderTargetInput`` lets you read _that_ back.

Reply:
Ahh ok